src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/PkcsPal.cs

```diff
@@ -0,0 +1,98 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Text;
+using System.Diagnostics;
+using System.Security.Cryptography;
+using System.Security.Cryptography.Pkcs;
+using System.Security.Cryptography.X509Certificates;
+
+namespace Internal.Cryptography
+{
+    internal abstract partial class PkcsPal
+    {
+        protected PkcsPal()
+        {
+        }
+
+        /// <summary>
+        /// Encrypt and encode a CMS. Return value is the RFC-compliant representation of the CMS that can be transmitted "on the wire."
+        /// </summary>
+        public abstract byte[] Encrypt(CmsRecipientCollection recipients, ContentInfo contentInfo, AlgorithmIdentifier contentEncryptionAlgorithm, X509Certificate2Collection originatorCerts, CryptographicAttributeObjectCollection unprotectedAttributes);
+
+        /// <summary>
+        /// Decode an encoded CMS.
+        ///    Call RecipientInfos on the returned pal object to get the recipients.
+        ///    Call TryDecrypt() on the returned pal object to attempt a decrypt for a single recipient.
+        /// </summary>
+        public abstract DecryptorPal Decode(byte[] encodedMessage, out int version, out ContentInfo contentInfo, out AlgorithmIdentifier contentEncryptionAlgorithm, out X509Certificate2Collection originatorCerts, out CryptographicAttributeObjectCollection unprotectedAttributes);
+
+        // 
+        // Encodes and decoders for the various Pkcs9 attribute types.
+        //
+
+        public abstract byte[] EncodeDocumentDescription(string documentDescription);
+        public abstract string DecodeDocumentDescription(byte[] encodedDocumentDescription);
+
+        public abstract byte[] EncodeDocumentName(string documentName);
+        public abstract string DecodeDocumentName(byte[] encodedDocumentName);
+
+        public abstract byte[] EncodeSigningTime(DateTime signingTime);
+        public abstract DateTime DecodeSigningTime(byte[] encodedSigningTime);
+
+        public abstract string DecodeContentType(byte[] encodedContentType);
+
+        public abstract byte[] DecodeMessageDigest(byte[] encodedMessageDigest);
+
+        /// <summary>
+        /// Implements the ContentInfo.GetContentType() behavior.
+        /// </summary>
+        public abstract Oid GetEncodedMessageType(byte[] encodedMessage);
+
+        /// <summary>
+        /// EnvelopedCms.Decrypt() looks for qualifying certs from the "MY" store (as well as any "extraStore" passed to Decrypt()).
+        /// This method encapsulates exactly what a particular OS considers to be "the MY store."
+        /// </summary>
+        public abstract void AddCertsFromStoreForDecryption(X509Certificate2Collection certs);
```

Comment:
We were debating whether or not to carry that mandatory pull-from-My behavior forward on Unix, so I wanted to leave that hook.

If it turns out the Unix implementation of that is “just do what Windows does”, we can easily remove that.